projects/web-wallet/rpc-protocol.md

```diff
@@ -0,0 +1,330 @@
+7. `onlyAvailableUtxos` - (Defaults to true) Get only available `utxos`, ignoring locked ones.
+
+**Example Request:**
+
+```json
+{
+  "id": 3,
+  "jsonrpc": "2.0",
+  "method": "htr_getUtxos",
+  "params": {
+	"maxUtxos": 50,
+	"token": "BTR",
+	"amount": 1
+  }
+}
+```
+
+**Example Response:**
+
+```json
+{
+  "total_amount_available": 5,
+  "total_utxos_available": 10,
+  "total_amount_locked": 0,
+  "total_utxos_locked": 0,
+  "utxos": [
+    {
+      "address": "HNnK9wgUVL6Cjzs1K3jpoGgqQTXCqpAnW8",
+      "amount": 1,
+      "tx_id": "00fff7a3c6eb95ec3343bffcfca9a3a0d3e243462ae7de1f200cdd76716140fb",
+      "locked": false,
+      "index": 0
+    }
+  ]
+}
+```
+
+
+### 2.1.4 `htr_signWithAddress`
+
+The `htr_signWithAddress` method requests a signed message using an address' private key that can be verified.
+
+**Parameters**
+
+1. `message` - String containing a message to be signed
+2. `addressIndex` - Address index to sign the message with
+
+**Request:**
+
+```json
+{
+  "id": 3,
+  "jsonrpc": "2.0",
+  "method": "htr_signWithAddress",
+  "params": {
+    "message": "sign-me",
+    "address": 5
+  }
+}
+```
+**Response:**
+
+```json
+{
+  "message": "H2/QA48wJhvbKJH6hWXUJPHK5kUTwx4XAxuZGFzUeSPwLeFBv206lOs8ECHMTV/KilGbFP1p5e52mMGTUDFT7WM="
+}
+```
+
+### 2.1.5 `htr_getBalance`
+
+Different from account-based blockchains where users are used to always using the same address on each transaction, getting the balance on hathor involves going through a list of addresses from the user's wallet and calculating the sum of all utxos.
+
+So in order for a `dApp` to know the balance of a given token for the connected user's wallet, we must have a method to fetch it.
+
+The `htr_getBalance` method fetches the balance for a given token.
+
+**Parameters**
+
+1. `token` - (Defaults to `HTR`) May be specified for a token other than `HTR`
+
+**Request:**
+
+```json
+{
+  "id": 4,
+  "jsonrpc": "2.0",
+  "method": "htr_getBalance",
+  "params": {
+    "token": "000023a025d513c5c75410e771c297b23edc88b8fac7da1eff6fdc81e628c20d"
+  }
+}
+```
+**Response:**
+
+```json
+{
+  "id": 4,
+  "jsonrpc": "2.0",
+  "result": {
+    "available": 500,
+    "locked": 0
+  }
+}
+```
+
+
+## 2.1.5 Error codes
+
+We should follow the [JSON-RPC](https://www.jsonrpc.org/specification#error_object) specification for error codes, so errors generated by the "server", which in our case, is the wallet handling the requests should be in the `-32099` to `-32000` range.
+
+| Error | Error Type | Description                        |
+|------------|------------------------------------|---|
+| -32050     | WALLET_NOT_READY | Wallet is not ready yet              |
+| -32051     | REQUEST_REJECTED | Request rejected by the user |
+| -32052     | MISSING_FUNDS | Missing funds to fulfill transaction |
+| -32053 | INVALID_PARAMETERS | Invalid parameters |
+| -32099     | GENERIC_FAILURE | Generic failure |
```

Comment:
Maybe UNKNOWN FAILURE or UNEXPECTED FAILURE?

Reply:
Yes! Done in 5e1d1efb147d1c61347842219bfb542f61a39f64

Reply:
Refactored it to `UNEXPECTED_FAILURE` in 0b6a49e2be9bb6483ce5494fdf1f1209ba3f4cfb

Reply:
Thanks!